File: java/src/main/java/org/urlshortener/dal/repositories/impl/ShortUrlRepositoryImpl.java
package org.urlshortener.dal.repositories.impl;

import lombok.NonNull;
import org.urlshortener.dal.entities.ShortUrl;
import org.urlshortener.dal.repositories.ShortUrlRepository;

import javax.enterprise.context.RequestScoped;
import javax.persistence.EntityManager;
import javax.persistence.PersistenceContext;
import java.util.List;

@RequestScoped
public class ShortUrlRepositoryImpl implements ShortUrlRepository {

    @PersistenceContext
    private final EntityManager em;

    public ShortUrlRepositoryImpl(EntityManager entityManager) {this.em = entityManager;}

    @Override
    public ShortUrl merge(@NonNull ShortUrl shortUrl) {
        return this.em.merge(shortUrl);
    }

    @Override
    public void deleteById(@NonNull Long id, @NonNull String userId) {
        var query = this.em.createQuery("delete from ShortUrl s where s.id = :id and s.userId = :userId");
        query.setParameter("id", id);
        query.setParameter("userId", userId);
        query.executeUpdate();
    }

    @Override
    public Long countByShortName(@NonNull String shortName) {
        var query = this.em
                .createQuery("select count(s) from ShortUrl s where s.shortName = :shortName", Long.class);
        query.setParameter("shortName", shortName);
        return query.getSingleResult();
    }

    @Override
    public ShortUrl findByShortName(@NonNull String shortName) {
        var query = this.em
                .createQuery("select s from ShortUrl s where s.shortName = :shortName", ShortUrl.class);
        query.setParameter("shortName", shortName);
        return query.getSingleResult();
    }

    @Override
    public ShortUrl findByIdAndUserId(Long id, String userId) {
        var query = this.em
                .createQuery("select s from ShortUrl s where s.id = :id and s.userId = :userid", ShortUrl.class);
        query.setParameter("id", id);
        query.setParameter("userid", userId);
        return query.getSingleResult();
    }

    @Override
    public boolean existsByIdAndUserId(Long id, String userId) {
        var query = this.em
                .createQuery("select count(s) from ShortUrl s where s.id = :id and s.userId = :userid", Long.class);
        query.setParameter("id", id);
        query.setParameter("userid", userId);
        return query.getSingleResult() != 0;
    }

    @Override
    public List<ShortUrl> findAllByUserId(String userId) {
        var query =
                this.em.createQuery("select s from ShortUrl s where s.userId = :userid", ShortUrl.class);
        query.setParameter("userid", userId);
        return query.getResultList();
    }
}

File: src/main/kotlin/com/urlshortener/dal/repositories/impl/ShortUrlRepositoryImpl.kt
@file:Suppress("PLATFORM_CLASS_MAPPED_TO_KOTLIN")

package com.urlshortener.dal.repositories.impl

import com.urlshortener.dal.entities.ShortUrl
import com.urlshortener.dal.repositories.ShortUrlRepository
import javax.enterprise.context.RequestScoped
import javax.persistence.EntityManager
import javax.persistence.PersistenceContext
import java.lang.Long as JavaLong

@RequestScoped

class ShortUrlRepositoryImpl(@PersistenceContext private val em: EntityManager) : ShortUrlRepository {
    override fun merge(shortUrl: ShortUrl): ShortUrl = em.merge(shortUrl)

    override fun deleteById(id: Long) {
    }

    //todo unique constraint????
    override fun countByShortName(shortName: String): Long =
        em.createQuery("select count(s) from ShortUrl s where s.shortName = :shortName", JavaLong::class.java).run {
            setParameter("shortName", shortName)
            singleResult.toLong()
        }

    override fun findByShortName(shortName: String): ShortUrl? =
        em.createQuery("select s from ShortUrl s where s.shortName = :shortName", ShortUrl::class.java).run {
            setParameter("shortName", shortName)
            singleResult
        }

    override fun findByIdAndUserId(id: Long, userId: String): ShortUrl? =
        em.createQuery("select s from ShortUrl s where s.id = :id and s.userId = :userid", ShortUrl::class.java).run {
            setParameter("id", id)
            setParameter("userid", userId)
            singleResult
        }

    override fun existsByIdAndUserId(id: Long, userId: String): Boolean =
        em.createQuery("select count(s) from ShortUrl s where s.id = :id and s.userId = :userid",JavaLong::class.java)
            .run {
                setParameter("id", id)
                setParameter("userid", userId)
                singleResult.toLong() != 0L
            }

    override fun findAllByUserId(userId: String): List<ShortUrl> =
        em.createQuery("select s from ShortUrl s where s.userId = :userid", ShortUrl::class.java).run {
            setParameter("userid", userId)
            resultList.toList()
        }
}